testsuite/tests/typing-poly/poly_params.ml

```diff
@@ -0,0 +1,202 @@
+(* TEST
+   * expect
+*)
+
+let poly1 (id : 'a. 'a -> 'a) = id 3, id "three"
+[%%expect {|
+val poly1 : ('a. 'a -> 'a) -> int * string = <fun>
+|}];;
+
+let _ = poly1 (fun x -> x)
+[%%expect {|
+- : int * string = (3, "three")
+|}];;
+
+let _ = poly1 (fun x -> x + 1)
+[%%expect {|
+Line 1, characters 14-30:
+1 | let _ = poly1 (fun x -> x + 1)
+                  ^^^^^^^^^^^^^^^^
+Error: This argument has type int -> int which is less general than
+         'a. 'a -> 'a
+|}];;
+
+let id x = x
+let _ = poly1 id
+[%%expect {|
+val id : 'a -> 'a = <fun>
+- : int * string = (3, "three")
+|}];;
+
+let _ = poly1 (id (fun x -> x))
+[%%expect {|
+Line 1, characters 14-31:
+1 | let _ = poly1 (id (fun x -> x))
+                  ^^^^^^^^^^^^^^^^^
+Error: This argument has type 'b -> 'b which is less general than
+         'a. 'a -> 'a
+|}];;
+
+let poly2 : ('a. 'a -> 'a) -> int * string =
+  fun id -> id 3, id "three"
+[%%expect {|
+val poly2 : ('a. 'a -> 'a) -> int * string = <fun>
+|}];;
+
+let _ = poly2 (fun x -> x)
+[%%expect {|
+- : int * string = (3, "three")
+|}];;
+
+let _ = poly2 (fun x -> x + 1)
+[%%expect {|
+Line 1, characters 14-30:
+1 | let _ = poly2 (fun x -> x + 1)
+                  ^^^^^^^^^^^^^^^^
+Error: This argument has type int -> int which is less general than
+         'a. 'a -> 'a
+|}];;
+
+let poly3 : 'b. ('a. 'a -> 'a) -> 'b -> 'b * 'b option =
+  fun id x -> id x, id (Some x)
+[%%expect {|
+val poly3 : ('a. 'a -> 'a) -> 'b -> 'b * 'b option = <fun>
+|}];;
+
+let _ = poly3 (fun x -> x) 8
+[%%expect {|
+- : int * int option = (8, Some 8)
+|}];;
+
+let _ = poly3 (fun x -> x + 1) 8
+[%%expect {|
+Line 1, characters 14-30:
+1 | let _ = poly3 (fun x -> x + 1) 8
+                  ^^^^^^^^^^^^^^^^
+Error: This argument has type int -> int which is less general than
+         'a. 'a -> 'a
+|}];;
+
+let rec poly4 p (id : 'a. 'a -> 'a) =
+  if p then poly4 false id else id 4, id "four"
+[%%expect {|
+val poly4 : bool -> ('a. 'a -> 'a) -> int * string = <fun>
+|}];;
+
+let _ = poly4 true (fun x -> x)
+[%%expect {|
+- : int * string = (4, "four")
+|}];;
+
+let _ = poly4 true (fun x -> x + 1)
+[%%expect {|
+Line 1, characters 19-35:
+1 | let _ = poly4 true (fun x -> x + 1)
+                       ^^^^^^^^^^^^^^^^
+Error: This argument has type int -> int which is less general than
+         'a. 'a -> 'a
+|}];;
+
+let rec poly5 : bool -> ('a. 'a -> 'a) -> int * string =
+  fun p id ->
+    if p then poly5 false id else id 5, id "five"
+[%%expect {|
+val poly5 : bool -> ('a. 'a -> 'a) -> int * string = <fun>
+|}];;
+
+let _ = poly5 true (fun x -> x)
+[%%expect {|
+- : int * string = (5, "five")
+|}];;
+
+let _ = poly5 true (fun x -> x + 1)
+[%%expect {|
+Line 1, characters 19-35:
+1 | let _ = poly5 true (fun x -> x + 1)
+                       ^^^^^^^^^^^^^^^^
+Error: This argument has type int -> int which is less general than
+         'a. 'a -> 'a
+|}];;
+
```

Comment:
I think we're missing some basic scope-escape /  value-restriction tests here, e.g.:
```ocaml
let escape f = poly5 true (fun x -> f x; x)
let nonval = poly5 true (let r = ref None in fun x -> r := Some x; x)
```

(It looks like the relevant checks are correct in the typechecker, but we need tests too)

Reply:
The value restriction one was already being tested by the `id (fun x -> x)` test, but this one is clearer because it's actually unsound so I've added it too. I've also added the scope-escape one.